subprojects/core/src/main/java/org/gradle/api/internal/DefaultMutationGuard.java

```diff
@@ -28,20 +28,28 @@ protected Boolean initialValue() {
 
     @Override
     public boolean isMutationAllowed() {
-        return mutationGuardState.get();
+        boolean mutationAllowed = mutationGuardState.get();
+        if (mutationAllowed) {
+            mutationGuardState.remove();
+        }
+        return mutationAllowed;
```

Comment:
Let's add some comments about why we are adding the extra complexity here.

Reply:
I added some comments.